core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/evaluationsteps/IntersectionQueryEvaluationStep.java

```diff
@@ -21,19 +22,19 @@
  */
 public class IntersectionQueryEvaluationStep implements QueryEvaluationStep {
 
-	private final QueryEvaluationStep leftArgDelayed;
-	private final QueryEvaluationStep rightArgDelayed;
+	private final QueryEvaluationStep leftArg;
+	private final Function<BindingSet, DelayedEvaluationIteration> rightArgDelayed;
 	private final Supplier<Set<BindingSet>> setMaker;
 
 	public IntersectionQueryEvaluationStep(QueryEvaluationStep leftArg, QueryEvaluationStep rightArg,
 			Supplier<Set<BindingSet>> setMaker) {
 		this.setMaker = setMaker;
-		leftArgDelayed = bs -> new QueryEvaluationStep.DelayedEvaluationIteration(leftArg, bs);
-		rightArgDelayed = bs -> new QueryEvaluationStep.DelayedEvaluationIteration(rightArg, bs);
+		this.leftArg = leftArg;
+		rightArgDelayed = bs -> new DelayedEvaluationIteration(rightArg, bs);
 	}
 
 	@Override
 	public CloseableIteration<BindingSet, QueryEvaluationException> evaluate(BindingSet bs) {
-		return new IntersectIteration<>(leftArgDelayed.evaluate(bs), rightArgDelayed.evaluate(bs), setMaker);
+		return new IntersectIteration<>(leftArg.evaluate(bs), rightArgDelayed.apply(bs), setMaker);
```

Comment:
The DelayedEvaluationIteration is hard for the compiler to optimize because it struggles to infer what kind of iteration will be produced and can't inline code as efficiently. The leftArg will anyways always be evaluated because the rightArg can only be safely ignored if the leftArg is empty.